src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs

```diff
@@ -90,25 +97,89 @@ public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache? lengthC
             lengthCache.Set(0);
 
             // number of fields + (type oid + field length) * member count
-            var length = sizeof(int) + sizeof(int) * 2 * _memberHandlers.Length;
-            foreach (var member in _memberHandlers)
+            var memberHandlers = _memberHandlers!;
+            var length = sizeof(int) + sizeof(int) * 2 * memberHandlers.Length;
+            foreach (var member in memberHandlers)
                 length += member.ValidateAndGetLength(value, ref lengthCache);
 
             return lengthCache.Lengths[position] = length;
         }
 
-        CompositeMemberHandler<T>[] CreateMemberHandlers() => CreateMemberHandlers((PostgresCompositeType)PostgresType, _typeMapper, _nameTranslator);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        void Initialize()
+        {
+            if (_memberHandlers is null)
+                InitializeCore();
+        }
 
-        static CompositeMemberHandler<T>[] CreateMemberHandlers(PostgresCompositeType pgType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
+        void InitializeCore()
         {
-            var clrType = typeof(T);
+            var pgType = (PostgresCompositeType)PostgresType;
```

Comment:
Change PostgresType's to PostgresCompositeType (with new)

Reply:
Unless I'm mistaken this property is used only once, so it isn't worth.

Reply:
Up to you, it's almost the same, and it means that validation that the type is composite happens earlier.

Reply:
It happens in `CompositeTypeHandlerFactory<T>`. By the way, a property is a pair of methods which should be called first, so it won't help as you suggested (: